www/docs/getting-started/upgrade-to-v4.md

```diff
@@ -0,0 +1,188 @@
+
+### 4. Named Parameters
+
+We have changed the arguments to our callbacks to the named parameters pattern. This way you don't have to use dummy `_` placeholders or other tricks.
+
+The signatures for the callback methods now look like this:
+
+```diff
+- signIn(user, account, profileOrEmailOrCredentials)
++ signIn({ user, account, profile, email, credentials })
+```
+
+```diff
+- redirect(url, baseUrl)
++ redirect({ url, baseUrl })
+```
+
+```diff
+- session(session, tokenOrUser)
++ session({ session, token, user })
+```
+
+```diff
+- jwt(token, user, account, OAuthProfile, isNewUser)
++ jwt({ token, user, account, profile, isNewUser })
+```
+
+Introduced in https://github.com/nextauthjs/next-auth/releases/tag/v4.0.0-next.17
+
+### 5. useSession Hook
+
+The `useSession` hook has been updated to return an object. This allows you to test states much more cleanly with the new `status` option.
+
+```diff
+- const [ session, loading ] = useSession()
++ const { data: session, status } = useSession()
++ const loading = status === "loading"
+```
+
+[Check the docs](https://next-auth.js.org/getting-started/client#usesession) for the possible values of both `session.status` and `session.data`.
+
+Introduced in https://github.com/nextauthjs/next-auth/releases/tag/v4.0.0-next.18
+
+### 6. nodemailer
+
+[`nodemailer`](https://npmjs.com/package/nodemailer) is no longer a dependency and added by default. If you are using the Email provider you can install it in your project manually, or use any other Email library in the [`sendVerificationRequest`](/configuration/providers#options-1#:~:text=sendVerificationRequest) callback.
+
+Introduced in https://github.com/nextauthjs/next-auth/releases/tag/v4.0.0-next.2
+
+### 7. Logger API
+
+The logger API has been simplified to use at most two parameters, where the second is usually an object (`metadata`) containing an `error` object. If you are not using the logger settings you can ignore this change.
+
+```diff
+// [...nextauth.js]
+import log from "some-logger-service"
+...
+logger: {
+- error(code, ...message) {},
++ error(code, metadata) {},
+- warn(code, ...message) {},
++ warn(code) {}
+- debug(code, ...message) {}
++ debug(code, metadata) {}
+}
+```
+
+Introduced in https://github.com/nextauthjs/next-auth/releases/tag/v4.0.0-next.19
+
+### 8. Events
+
+Two event signatures changes to also use the named parameters pattern, `signOut` and `updateUser`.
+
+```diff
+// [...nextauth].js
+...
+events: {
+- signOut(tokenOrSession),
++ signOut({ token, session }), // token if using JWT, session if DB persisted sessions.
+- updateUser(user)
++ updateUser({ user })
+}
+```
+
+Introduced in https://github.com/nextauthjs/next-auth/releases/tag/v4.0.0-next.20
+
+### 9. Providers
+
+All OAuth providers' `profile` callback are expected to only return these fields by default from now on: `id`, `name`, `email`, and `image` at most. If any of these are missing values, they should be set to `null`.
```

Comment:
Should point out that this is only the case when someone wants to add their provider to the built-in ones. The reason for this BTW, because I intend to require only these fileds on default user models for the official adapters as well.